### PR TITLE
button timeline working with Date objects

### LINKED ISSUE #35: button timeline does not work with Date objects
### Expected Behavior
I would expect to be able to pass fully-validated JavaScript Date objects to the various methods.

### Current Behavior
Setting the `domain`, `ticks`, in addition to the current `selection` causes bugs.

### Steps to Reproduce (for bugs)
http://jsfiddle.net/xs5p2ujh/1/

### Your Environment
* Browser Name: Chrome
* Operating System and Version: macOS High Sierra

### PR DESCRIPTION
**Pull request by cnavarreteliz**
closes #35 

### Description

I solved this bug adding a test that checks if the object is Date.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [ ] I have documented all new methods.
- [ ] I have written tests for all new methods/functionality.
- [ ] I have written examples for all new methods/functionality.

### CODE REVIEW

**Review by davelandry — CHANGES_REQUESTED** [review]
This solution will only work if the timeline is for years, but will break for thinks like month and day intervals. Have you tried simply setting `this._tickFormat` based on the variable you use for the tick widths?

File: src/Timeline.js [hunk 1]
```diff
@@ -237,11 +247,13 @@ export default class Timeline extends Axis {
   render(callback) {
     const {height, y} = this._position;
 
+    if (this._ticks) this._ticks = this._ticks.map(this._dateToString);
```
Comment (davelandry) [inline src/Timeline.js]: I don't understand this logic. 2 lines below this you transform the ticks back into dates anyways and create a tickFormat variable there.

**Review by davelandry — CHANGES_REQUESTED** [review]

File: src/Timeline.js (hunk 1 above)
Comment (cnavarreteliz) [reply src/Timeline.js]: `let ticks = this._ticks ? this._ticks.map(date) : this._domain.map(date);` and `tickFormat` they're for to calculate internally the width of the buttons.

The goal of `if (this._ticks) this._ticks = this._ticks.map(this._dateToString);` is to standardize the ticks passed by the user.

Comment (davelandry) [reply src/Timeline.js]: Right, but `tickFormat` you create in `_dateToString` does not use the domain and range, so it's tick formatter is actually different than the one you use for the width. The d3 time formatter takes the available space and domain of the data into account when suggesting a formatter.

I did a little investigation, and simply adding this one line to line #244 solve the bug for me:

```js
if (!this._tickFormat) this._tickFormat = tickFormat;
```

**Review by davelandry — CHANGES_REQUESTED** [review]

**Comment by davelandry** [comment]
The issue is happening because internally in d3plus-axis it knows to use a date formatter if the scale type if "time", but in our case of buttons we need to use "ordinal".

**Review by davelandry — CHANGES_REQUESTED** [review]
I'm getting an error when using this latest code:

```js
function toDate(d) {
  const date = new Date(`${d}/01/01`);
  date.setFullYear(d);
  return date;
}

const timeline = new d3plus.Timeline()
  .domain([2001, 2004].map(toDate))
  .ticks([2001, 2002, 2003, 2004].map(toDate))
  .render();
```

**Comment by cnavarreteliz** [comment]
@davelandry I'm working in this bug. It's caused here: `Array.from(Array(this._domain[1] - this._domain[0] + 1), (_, x) => this._domain[0] + x)` when you pass a Date in the domain. But the rest of the simplification in the code it's working fine.

**Review by davelandry — APPROVED** [review]

**Comment by davelandry** [comment]
@cnavarreteliz can you resolve the conflicts?

**Comment by cnavarreteliz** [comment]
@davelandry done! This is one of the most happiest days in my life of programmer haha and I think that I learnt so much Javascript with d3plus-timeline.

**Pull request merged**